### ISSUE TITLE
format:iife's output is inconsistent

### ISSUE THREAD

**Issue by bfred-it** [rollup/rollup#827]
`iife` sometimes overrides the chosen `moduleName`, sometimes extends it.
## Override ([rollupjs.org demo](https://rollupjs.org/repl?version=0.43.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTdEJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmlpZmUlMjIlMkMlMjJtb2R1bGVOYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIybW9kdWxlSWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==))

``` js
export default function () {};
```

``` js
var myBundle = (function () {
    'use strict';
    function main () {};
    return main;
}());
```
## Extend ([rollupjs.org demo](https://rollupjs.org/repl?version=0.43.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZm9vJTIwJTNEJTIwMSUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJpaWZlJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMm1vZHVsZUlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=))

``` js
export const foo = 1;
```

``` js
(function (exports) {
    'use strict';
    const foo = 1;
    exports.foo = foo;
}((this.myBundle = this.myBundle || {})));
```

**Comment by TrySound** [comment]
This is the difference between default and names exports. If you have both then module is extended with `default` property.

**Comment by bfred-it** [comment]
I see that but it's still inconsistent:

## Override

```html
<script>
window.MyBundle = 'some existing variable';
</script>
<script src='rollup-output.js'></script>
<script>
    window.myBundle === function main () {} // true
</script>
```

## Extend

```html
<script>
window.MyBundle = 'some existing variable';
</script>
<script src='rollup-output.js'></script>
<script>
    window.myBundle === 'some existing variable' // true
    window.myBundle.foo === 1; // FALSE
</script>
```

**Comment by TrySound** [comment]
Do you mean also return exports?

**Comment by bfred-it** [comment]
Yes, sorry, edited

**Comment by TrySound** [comment]
@bfred-it Extends is required behavior. For example d3 use it. It's splitted in many modules since v4. But they all refers to d3 global variable. If we will do "consistent" like you want. Each module will replace previous. It's not good.

**Comment by bfred-it** [comment]
That doesn't look like it "correct" either: why does `[d3-ease](https://github.com/d3/d3-ease)` use `window.d3`? It should export to `window.d3ease`. Its current setup seem to favor that specific situation, but it doesn't seem right. `d3-ease` exports `d3-ease`, not `d3`. This "plugin-like" (or "extend") export should be its own type probably

**Comment by mbostock** [comment]
D3’s approach is designed to favor backwards compatibility with D3 3.x, and
more so, to favor compatibility in a vanilla environment between using the
default bundle (d3/d3) and using separate modules (multiple script
tags)—you can write code the same way, and you can even cat the scripts
together to hand roll a custom bundle.

It also means that the documentation is simpler: I can document
d3.easeCubic and it is the same regardless of environment, which is
critical for helping new programmers that have not mastered the fragmented
convolutions of modern JavaScript development.

Lastly it means that the D3 namespace is shared across all D3 modules,
which encourages people to name symbols in a way that will not conflict,
which is nice even if you are using ES2015 modules in your application.

**Comment by bfred-it** [comment]
I understand that but the whole setup depends on "d3-ease" having a "d3" `moduleName`. Nobody sees a problem with that? It feels like a hack (besides the issue already shown above) and should probably be `--extends d3` (since that's what it does)

**Comment by Victorystick** [comment]
Extending an existing namespace was accepted as an enhancement by #493 and soon after became a feature. I can sort of see the use case, but it's a problem that doesn't really follow any specification and could be solved better in some other way.

**Comment by Andarist** [comment]
@bfred-it 
The closed issue https://github.com/rollup/rollup/issues/1397 is not exactly about the same thing, however they relate to each other closely.

Is there anything I can do to move those forward? Implementation should be rather trivial. Only the API needs to be established.

**Comment by bfred-it** [comment]
How does that dffer from this one?

**Comment by Andarist** [comment]
This issue is focused on inconsistent behaviour of overriding vs extending the configured `moduleName`.

My issue was rather focused on omitting `this` at all. I want to simply configure the exported object to be assigned to a local var.

**Comment by bfred-it** [comment]
Whatever the explanation, the intent is the same: avoiding `this`.

I'd go for moving the current behavior from `--moduleName` to `--extends`, and your suggested output as the default for `--moduleName` with multiple exports, as that would be consistent with the single export (both would start with `var moduleName = iife`

**Comment by Andarist** [comment]
Whats ur policy about breaking changes?

I'll try to prepare a PR for this in following days (hopefully till the end of the week).

**Comment by bfred-it** [comment]
(I'm not the maintainer)

Rollup is still 0.x, so every minor release is breaking and there have been 42(?) breaking releases so far 😃

**Comment by Victorystick** [comment]
@Andarist Making the output more consistent seems like a valid reason to change the behavior. I don't expect a lot of people to rely on this particular feature. Thoughts, @mbostock?

**Comment by bfred-it** [comment]
D3.js does, so as long as the current behavior is available under a different name it's just a matter of changing it in D3's build.

**Comment by mbostock** [comment]
This pattern is used by D3’s 30+ “official” modules and by D3 v4 plugins maintained by their respective authors, so I imagine changing the behavior would cause some confusion… but, that shouldn’t stop you from making a breaking change if you believe it’s the right way forward.

**Comment by bfred-it** [comment]
Perhaps we could include a temporary warning. It could target `--moduleName d3` or all multiple exports, to explain how the change may affect them. I can write this one the feature is in.

**Comment by Andarist** [comment]
Just to let you know - I've already started working on this feature. When I send a PR, I'll mention you all to establish the right migration path (and possible warning policy to be included in the source).

**Comment by Andarist** [comment]
@mbostock could u explain how do u use current behaviour in d3? Ive checked the packages and umd micro-libraries but I cannot find IIFE bundles anywhere there.

I would like to reassure I wont break anything for you before finishing my PR (just 1 failing test to go).

**Comment by bfred-it** [comment]
`window.d3` is modular and as you load multiple plugins with the same `moduleName` they will just be added as properties to `window.d3`:

```js
window.d3.moduleOne
window.d3.moduleTwo
Etc
```

**Comment by mbostock** [comment]
@Andarist Right, none of the D3 modules or plugins use iife; they use the umd output format.

**Comment by bfred-it** [comment]
Yeah, it looks like this doesn't only affect IIFE but also UMD (which is essentially an IIFE plus CJS and AMD support) so you'll find the same `global.d3 = global.d3 || {}` code in the official d3 build as well.

**Comment by Rich-Harris** [comment]
Released 0.44 with this change
